third_party/move/move-compiler-v2/src/flow_insensitive_checkers.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Flow-insensitive checks can be done on the AST.
+//!
+//! Warnings about Unused parameter and local variable
+//!   "Unused assignment or binding for local 's'. Consider removing, replacing with '_' or prefixing with '_' (e.g., '_r_ref')
+
+use codespan_reporting::diagnostic::Severity;
+use move_model::{
+    ast::{ExpData, TempIndex},
+    model::{GlobalEnv, Loc, NodeId, Parameter},
+    symbol::Symbol,
+};
+use std::{collections::BTreeSet, iter::Iterator};
+
+/// Warns about all parameters and local variables that are unused.
+pub fn check_for_unused_vars_and_params(env: &mut GlobalEnv) {
+    for module in env.get_modules() {
+        for func in module.get_functions() {
+            if let Some(def) = &*func.get_def() {
+                let params = &func.get_parameters();
+                find_unused_params_and_vars(env, params, def)
+            }
+        }
+    }
+}
+
+fn find_unused_params_and_vars(env: &GlobalEnv, params: &[Parameter], exp: &ExpData) {
+    let mut visitor = SymbolVisitor::new(env, params);
+    exp.visit_pre_post(&mut |post, exp_data| visitor.entry(post, exp_data));
+    visitor.check_parameter_usage();
+}
+
+/// Tracks things of type `V` which are visible from below in a tree, such as free/used variables,
+/// etc.  `values` is cleared when entering a scope, but the old value saved so it can be re-added
+/// once the scope is finished.
+struct ScopedVisibleSet<V> {
+    saved: Vec<BTreeSet<V>>,
+    values: BTreeSet<V>,
+}
+
+impl<V> ScopedVisibleSet<V>
+where
+    V: Ord + Copy,
+{
+    pub fn new() -> Self {
+        Self {
+            saved: Vec::new(),
+            values: BTreeSet::new(),
+        }
+    }
+
+    #[allow(unused)]
+    /// Clear everything for reuse on a new function.
```

Comment:
As this is a private struct and not a public API, perhaps someone can just add it once it is really needed.

Reply:
Ok, I was keeping it until I was sure i didn't need it.